File: 第五学期/密码学/汇报/otherRC6/RC6/RC6.cpp
#include "RC6.hpp"

RC6::RC6(unsigned int W, unsigned int R, unsigned int B){
  w = W;
  r = R;
  b = B;
  log_w = (unsigned int)log2(w);
  modulo = std::pow(2, w);
  S = new unsigned int[2 * r + 4];
}

void RC6::rc_constraints(const unsigned int &w, unsigned int &p, unsigned int &q){
  p = (unsigned int)std::ceil(((M_E - 2) * std::pow(2, w)));
  q = (unsigned int)((1.618033988749895 - 1) * std::pow(2, w));    // Golden Ratio
}

/******************************************************************
 * Function: left_rot
 * Input: <int> a, <unsigned int> b, <unsigned int> w
 * Output: int
 * Description: Rotate the w-bit word a to the left by the amount
 *              given by the least significant log w bits of b
 ******************************************************************/
int RC6::left_rot(unsigned int a, unsigned int b, unsigned int w){
  b <<= w - log_w;
  b >>= w - log_w;
  return (a << b) | (a >> (w - b));  
}

/******************************************************************
 * Function: right_rot
 * Input: <int> a, <unsigned int> b, <unsigned int> w
 * Output: int
 * Description: Rotate the w-bit word a to the right by the amount
 *              given by the least significant log w bits of b
 ******************************************************************/
int RC6::right_rot(unsigned int a, unsigned int b, unsigned int w){
  b <<= w - log_w;
  b >>= w - log_w;
  return (a >> b) | (a << (w - b));
}

/******************************************************************
 * Function: little_endian
 * Input: <std::string>
 * Output: std::string
 * Description: Convert input string to little endian version by bytes
 *              and return new little endian string. Assuming
 *              each character represents a byte
 ******************************************************************/
std::string RC6::little_endian(std::string str){
  std::string endian;
  
  if(str.length() % 2 == 0){
    for(std::string::reverse_iterator r_it = str.rbegin();
	     r_it != str.rend();r_it = r_it + 2){
      endian.push_back(*(r_it+1));
      endian.push_back(*r_it);
    }
  }else{
    str = "0" + str;
    for(std::string::reverse_iterator r_it = str.rbegin();
      r_it != str.rend();r_it = r_it + 2){
      endian.push_back(*(r_it+1));
      endian.push_back(*r_it);
    }
  }

  return endian;
}

std::string RC6::hex_to_string(unsigned int A, unsigned int B, unsigned int C, unsigned int D){

  std::string strA, strB, strC, strD, result;

  std::stringstream ss;
  ss << std::setfill('0') << std::setw(4) <<std::hex << A;
  strA = little_endian(ss.str());
  ss.str("");
  ss.clear();

  ss << std::setfill('0') << std::setw(4) <<std::hex << B;
  strB = little_endian(ss.str());
  ss.str("");
  ss.clear();
  
  ss << std::setfill('0') << std::setw(4) <<std::hex << C;
  strC = little_endian(ss.str());
  ss.str("");
  ss.clear();

  ss << std::setfill('0') << std::setw(4) <<std::hex << D;
  strD = little_endian(ss.str());
  ss.str("");
  ss.clear();

  result = strA + strB + strC + strD;

  return result;

}

/******************************************************************
 * Function: key_schedule
 * Input: <std::string> key
 * Output: void
 * Description: Generates the key schedule for RC6. 
 *              The input takes the user-supplied b byte key preloaded
 *              in c-word L[0, ... , c-1]
 *              The function then outputs the w-bit round keys
 *              S[0, ... , 2r+3]
 ******************************************************************/
void RC6::key_schedule(std::string key){
  const unsigned int w_bytes = std::ceil((float)w / 8);
  const unsigned int c = std::ceil((float)b / w_bytes);

  unsigned int p, q;
  rc_constraints(w, p, q);

  L = new unsigned int[c];
  for(int i = 0; i < c; i++){
    L[i] = strtoul(little_endian(key.substr(w_bytes * 2 * i, w_bytes * 2)).c_str(), NULL, 16);
  }  

  S[0] = p;
  for(int i = 1; i <= (2 * r + 3); i++){
    S[i] = (S[i - 1] + q) % modulo;
  }

  unsigned int A = 0, B = 0, i = 0, j = 0;
  int v = 3 * std::max(c, (2 * r + 4));
  for(int s = 1; s <= v; s++){
    A = S[i] = left_rot((S[i] + A + B) % modulo, 3, w);
    B = L[j] = left_rot((L[j] + A + B) % modulo, (A + B), w);
    i = (i + 1) % (2 * r + 4);
    j = (j + 1) % c;
  }
}

/******************************************************************
 * Function: encrypt
 * Input: <const std::string>
 * Output: std::string
 * Description: Encrypt plaintext from the input string to ciphertext
 ******************************************************************/
std::string RC6::encrypt(const std::string &text){
  std::string result;
  
  unsigned int A, B, C, D;
  A = strtoul(little_endian(text.substr(0, 8)).c_str(), NULL, 16);
  B = strtoul(little_endian(text.substr(8, 8)).c_str(), NULL, 16);
  C = strtoul(little_endian(text.substr(16, 8)).c_str(), NULL, 16);
  D = strtoul(little_endian(text.substr(24, 8)).c_str(), NULL, 16);

  int32_t t, u, temp;

  B += S[0];
  D += S[1];
  for(int i = 1; i <= r; ++i){
    t = left_rot((B * (2 * B + 1)) % modulo, log_w, w);
    u = left_rot((D * (2 * D + 1)) % modulo, log_w, w);
    A = left_rot((A ^ t), u , w) + S[2 * i];
    C = left_rot((C ^ u), t, w) + S[2 * i + 1];
    temp = A;
    A = B;
    B = C;
    C = D;
    D = temp;
  }

  A += S[2 * r + 2];
  C += S[2 * r + 3];

  result = hex_to_string(A, B, C, D);

  return result;
}

/******************************************************************
 * Function: decrypt
 * Input: <const std::string>
 * Output: std::string
 * Description: decrypt ciphertext from the input string to plaintext
 ******************************************************************/
std::string RC6::decrypt(const std::string &text){
  std::string result;
  
  unsigned int A, B, C, D;
  A = strtoul(little_endian(text.substr(0, 8)).c_str(), NULL, 16);
  B = strtoul(little_endian(text.substr(8, 8)).c_str(), NULL, 16);
  C = strtoul(little_endian(text.substr(16, 8)).c_str(), NULL, 16);
  D = strtoul(little_endian(text.substr(24, 8)).c_str(), NULL, 16);

  unsigned int t, u, temp;  
  
  C -= S[2 * r + 3];
  A -= S[2 * r + 2];
  for(int i = r; i >= 1; --i){
    temp = D;
    D = C;
    C = B;
    B = A;
    A = temp;
    u = left_rot((D * (2 * D + 1)) % modulo, log_w, w);
    t = left_rot((B * (2 * B + 1)) % modulo, log_w, w);
    C = right_rot((C - S[2 * i + 1]) % modulo, t, w) ^ u;
    A = right_rot((A - S[2 * i]) % modulo, u, w) ^ t;
  }
  D -= S[1];
  B -= S[0];

  result = hex_to_string(A, B, C, D);

  return result;
}



std::string RC6::run(const std::string &mode, const std::string &text, const std::string &key){

  std::string result;

  key_schedule(key);
  
  if(mode.compare(0, strlen("Encryption"), "Encryption") == 0){
    std::string encryption = encrypt(text);
    for(std::string::iterator it = encryption.begin(); it != encryption.end();it = it + 2){
      result.push_back(*it);
      result.push_back(*(it+1));
      result = result + " ";
    }
  }else if(mode.compare(0, strlen("Decryption"), "Decryption") == 0){
    std::string decryption = decrypt(text);
    for(std::string::iterator it = decryption.begin(); it != decryption.end();it = it + 2){
      result.push_back(*it);
      result.push_back(*(it+1));
      result = result + " ";
    }
  }

  return result;
}

RC6::~RC6(){
  delete S;
}

File: 5_第五学期/密码学课程设计/05191643-许万鹏-密码学课程设计/block_cipher/rc6.py
from math import e
from scipy.constants import golden
from math import ceil, log2

w = 32
r = 20
b = 16
MOD = pow(2, 32)


def ROL(x, s):
    x &= 0xFFFFFFFF  # F^8==2^32
    s %= w
    res = ((x << s | x >> (w - s))) & 0xFFFFFFFF
    return res


def ROR(x, s):
    return ROL(x, w - s)


def odd(x):
    x = int(x)
    while x % 2 == 0:
        x += 1
    return x


# def k2l(key, u):
#     L = []
#     res = ''
#     for i in range(0, len(key)):
#         if i % u == 0 and i != 0:
#             L.append(res)
#             res = ''
#         res += bin(ord(key[i]))[2:].zfill(8)
#     L.append(res)
#     return L


def extend_key(key):
    P_w = odd((e - 2) * pow(2, w))
    Q_w = odd((golden - 1) * pow(2, w))
    u = w // 8  # 1word=8byte
    c = ceil(b / u)

    k = [ord(ch) for ch in key] + [0] * (b - len(key))
    L = []
    for i in range(0, len(k), u):
        L.append((k[i] << 24) + (k[i + 1] << 16) + (k[i + 2] << 8) + k[i + 3])

    S = [0] * (2 * r + 4)
    S[0] = P_w
    for i in range(1, 2 * r + 3 + 1):
        S[i] = (S[i - 1] + Q_w) % MOD
    A = B = i = j = 0
    v = 3 * max(c, 2 * r + 4)
    while v := v - 1:
        A = S[i] = ROL(S[i] + A + B, 3)
        B = L[j] = ROL(L[j] + A + B, A + B)
        i = (i + 1) % (2 * r + 4)
        j = (j + 1) % (c)
    return S


def encrypt(plaintext, k):
    ciphertext = ''
    S = extend_key(k)
    bin_plaintext = ''.join([bin(ord(ch))[2:].zfill(8) for ch in plaintext])
    bin_plaintext += '0' * (128 - len(bin_plaintext) % 128)
    for l in range(0, len(bin_plaintext), 128):
        A = int(bin_plaintext[l + 00:l + 32], 2)
        B = int(bin_plaintext[l + 32:l + 64], 2)
        C = int(bin_plaintext[l + 64:l + 96], 2)
        D = int(bin_plaintext[l + 96:l + 128], 2)

        B = (B + S[0]) % MOD
        D = (D + S[1]) % MOD
        for i in range(1, r + 1):
            t = ROL(B * (2 * B + 1), int(log2(w)))
            u = ROL(D * (2 * D + 1), int(log2(w)))
            A = (ROL(A ^ t, u) + S[2 * i]) % MOD
            C = (ROL(C ^ u, t) + S[2 * i + 1]) % MOD
            (A, B, C, D) = (B, C, D, A)
        A = (A + S[2 * r + 2]) % MOD
        C = (C + S[2 * r + 3]) % MOD

        ciphertext += f'{A:032b}{B:032b}{C:032b}{D:032b}'
    res = hex(int(ciphertext, 2))[2:]
    print(res)


def decrypt(ciphertext, k):
    S = extend_key(k)
    plaintext = ''
    for l in range(0, len(ciphertext), 128):
        A = int(ciphertext[l + 00:l + 32], 2)
        B = int(ciphertext[l + 32:l + 64], 2)
        C = int(ciphertext[l + 64:l + 96], 2)
        D = int(ciphertext[l + 96:l + 128], 2)
        C = (C - S[2 * r + 3]) % MOD
        A = (A - S[2 * r + 2]) % MOD
        for i in range(r, 1 - 1, -1):
            (A, B, C, D) = (D, A, B, C)
            u = ROL(D * (2 * D + 1), 5)
            t = ROL(B * (2 * B + 1), 5)
            C = ROR(C - S[2 * i + 1], t) ^ u
            A = ROR(A - S[2 * i], u) ^ t
        D = (D - S[1]) % MOD
        B = (B - S[0]) % MOD
        plaintext += f'{A:032b}{B:032b}{C:032b}{D:032b}'

    res = ''
    for i in range(0, len(plaintext), 8):
        res += chr(int(plaintext[i:i + 8], 2))
    print(res)


while True:
    type = input()
    if type == 'E':
        plaintext = input("请输入明文:\n")
        k = input("请输入密钥(小于32字节):\n")
        encrypt(plaintext, k)
    elif type == 'D':
        ciphertext = input("请输入密文:\n")
        ciphertext = ''.join([f'{int(ch, 16):04b}' for ch in ciphertext])
        k = input("请输入密钥(小于32字节):\n")
        decrypt(ciphertext, k)
